### DISCUSSION TITLE
How to max out a Pi Zero (w)

### DISCUSSION THREAD

**Discussion by spaudio** [probonopd/MiniDexed#516] [Q&A]
Hi, as the Zero 2 is almost impossible to buy anywhere but I do have a few Zero 1 waiting for a better job, how would I best approach running 2 instances of MiniDexed?
I understand very well that resources are limited but my hope is that by reducing sample rate or polyphony, I might be able to get two instances running without reverb.
Any help is highly appreciated. Thanks!

**Comment by probonopd** [comment]
So far we run just one Tone Generator and no effects on the RPi Zero. I am not sure whether there is enough computing power to run 2 Tone Generators. I don't think we have attempted this, so worth a try. With reducing polyphony to half (again -- it is already half compared to other models) it _might_ work, at least in theory, but we have never tried.

Here we define the number of Tone Generators for the RPi 1:

https://github.com/probonopd/MiniDexed/blob/87dc5ce43213ccf9a3a5813c72cbafcb0c4bed8b/src/config.h#L34-L36

Try setting this to `2`.

And here we define the polyphony for the RPi 1:

https://github.com/probonopd/MiniDexed/blob/87dc5ce43213ccf9a3a5813c72cbafcb0c4bed8b/src/config.h#L42-L44

Try setting this to `4` and then try to increase it from there until you run into issues.

**Reply by Banana71** [reply]
It would be great if someone actually tried it.  Unfortunately, I have no idea what steps I have to take to create a working version of miniDexed.  I hope that the RPi1 has enough power for 8 TGs with 2 voice polyphony.  (drum module)

**Reply by probonopd** [reply]
Build is running: https://github.com/probonopd/MiniDexed/pull/517

**Reply by probonopd** [reply]
Available for download in #517 but untested, feedback appreciated.

**Reply by Banana71** [reply]
TG 1 works with 4 voices polyphony and TG2 stays silent, no changes to volume, midichannel or anything else makes TG2 sound.

**Reply by probonopd** [reply]
Not sure why that is; maybe we need multicore for more than 1 TG?

**Reply by spaudio** [reply]
I haven't dug into the sources but aren't the TGs threads that are independent from hardware cores?
BTW thanks a lot @probonopd for making a dedicated build, I highly appreciate it!

**Reply by probonopd** [reply]
All I remember is that @rsta2 had carefully juggled with TGs and cores for the multicore RPi models... sorry I won't be of much help here.

**Reply by spaudio** [reply]
From the fact that multicore models run 2 TGs on core 1, I would hope that 2 TGs on a single core are possible.

The flag ARM_ALLOW_MULTI_CORE seems to be evaluated at multiple positions so maybe these have to be changed as well?

**Reply by diyelectromusic** [reply]
If you enable ProfileEnabled in minidexed.ini then it gives an indication of how busy things are - i.e. if it is stalling or coping well with generating the audio.  Part of the problem on a single-core setup is that it has to poll all the UI stuff, MIDI and the tone generators as part of the same thread of execution, which means eventually something will suffer!

To chew this over some more, see the following:
* Architecture diagram in the developer area of the wiki (bottom diagram of this: https://github.com/probonopd/MiniDexed/wiki/Development).
* The ChunkSize parameter - fiddling with this might help (again I forget the details - but I think this relates to how much audio is calculated in each parse).
* The SampleRate - a lower sample rate will reduce quality, but might allow more processing to happen.

I was going to allow the pairing of two Pi V1s with my messing around here: https://github.com/probonopd/MiniDexed/discussions/504 - I've not tested it yet, but I was thinking it could probably work with two 1TG Pi 1s as much as two 8TG Pi 2,3,4s which would allow two TGs and 16 TGs respectively...

Kevin

**Reply by diyelectromusic** [reply]
> TG 1 works with 4 voices polyphony and TG2 stays silent, no changes to volume, midichannel or anything else makes TG2 sound.

I'm pretty sure that the non-multi-core build will only cope with one TG in the ProcessSound routines in Minidexed.cpp...

Yes, in https://github.com/probonopd/MiniDexed/blob/main/src/minidexed.cpp#L959 it only ever calls TG[0]->GetSamples

See the architecture diagram in the developers section of the Wiki for more on the differences between multi-core and single-core operation.

Kevin

**Reply by probonopd** [reply]
Would it be possible/advisable to set `ARM_ALLOW_MULTI_CORE` for RPi 1 but set the number of cores to 1?

**Reply by diyelectromusic** [reply]
So in addition to the dual sound generation, it will need a mixing stage, which just doesn't happen at all at the moment for a single core system.

I could take a bit of a look - I'm a fan of making the most out of the Pi V1 (see my blog!), but I suspect that now the supply shortages are starting to ease up, it will be simpler to hold out for a Pi Zero V2 or Pi 2-3-4! :)

Kevin

**Reply by diyelectromusic** [reply]
> Would it be possible/advisable to set ARM_ALLOW_MULTI_CORE for RPi 1 but set the number of cores to 1?

I don't think this would work - it is as much a property for circle, library code and the rest of the build as for MiniDexed.  No, the right way would be to implement mixing of two TGs properly for a single core, but I suspect with all the UI/MIDI to sort through as well it will always be sub optimal...

I'd take a look, but I've just trashed my Linux VM by mistake so will have to rebuild it before I can do anymore on MiniDexed!  Good job everything I'm working on is already in branches in GitHub!! :)

Kevin

**Comment by Banana71** [comment]
> I could take a bit of a look - I'm a fan of making the most out of the Pi V1 (see my blog!), but I suspect that now the supply shortages are starting to ease up, it will be simpler to hold out for a Pi Zero V2 or Pi 2-3-4! :)
> 
I also don't know why I'm so fixated on the one RPi Zero I have.  I'm sure I'll find a good job for the little one.  Maybe  as a MIDI controller including arpegiator or simple sequencer.  
But what about the RPi 4?  Have there already been attempts to get 12 or 16 TGs running on it?

**Reply by probonopd** [reply]
> But what about the RPi 4? Have there already been attempts to get 12 or 16 TGs running on it?

Don't think anyone has ever tried.

For me personally, the killer use case would be Unison as described [here](https://github.com/probonopd/MiniDexed/issues/315).

**Comment by diyelectromusic** [comment]
> But what about the RPi 4? Have there already been attempts to get 12 or 16 TGs running on it?

Not on one, but I've 16 TGs running together on a pair of 3A+s... :)

But a build that splits 16 TGs over the cores of the Pi 4 should just be a case of changing the numbers in the build...  Maybe try with:
`static const unsigned TGsCore1 = 4;
static const unsigned TGsCore23 = 6;`

and just see how you go??

Oh, but you'd have to add in more menu items here too: https://github.com/probonopd/MiniDexed/blob/87dc5ce43213ccf9a3a5813c72cbafcb0c4bed8b/src/uimenu.cpp#L42

Kevin

**Reply by diyelectromusic** [reply]
I now have a build for a RPi 4 that will give 16 TGs.  More here: https://github.com/probonopd/MiniDexed/discussions/504#discussioncomment-6502626

Kevin

**Comment by spaudio** [comment]
It looks like Pi Zero 2 boards are slowly becoming available again (July 2023).
@diyelectromusic Would you mind sharing your VM after re-building it?
BTW many thanks for mentioning the architecture Wiki! It explains a lot.

**Reply by diyelectromusic** [reply]
It's just a standard Linux Ubuntu with the tool chain for building circle-based apps as I described here: https://diyelectromusic.wordpress.com/2021/10/08/bare-metal-raspberry-pi-synth/

Then with all the user-set up I need to link to GitHub.  In that regard, it is personal to me, but it is nothing special.

Apparently it is possible to build MiniDexed using a cloud-based setup, see: https://github.com/probonopd/MiniDexed/wiki/Development - but that isn't something I've done.

Kevin

**Reply by diyelectromusic** [reply]
For what it's worth, I've written up my "notes to self" on getting back up and running for MiniDexed development in an Ubuntu Virtual Machine here: https://diyelectromusic.wordpress.com/2023/07/21/rebuilding-my-ability-to-build-minidexed/

Kevin

**Comment by seymourgler** [comment]
I'm building Minidexed with a RPi Zero as well. I'm fine with one TG, and even low polyphony, in that case is it possible to activate effects?

**Reply by diyelectromusic** [reply]
It's not particularly trivial I'm afraid.  It's not a configuration option, as I expect you've seen, but to build it into the code means moving some of the effects code from the multi-core ProcessSound routine to the non-multi-core ProcessSound routine in minidexed.cpp and fiddling about with the sound buffers and mixing (there is no mixing, naturally, on the singe-core, single TG version).

It's not something I've looked at in detail, but I know there were some PRs kicking around to revamp the effects code at some point.

Kevin
